ARCHITECTURE.md

```diff
@@ -0,0 +1,265 @@
+  contains some recommendations on handling Refresh Tokens securely.
+
+#### Auth code flow
+
+![Module dependencies](./documentation/diagrams/auth_code_flow.svg)
+
+- Discovery: `packages/*/src/login/oidc/IssuerConfigFetcher.ts`
+- Registration: `packages/*/src/login/oidc/ClientRegistrar.ts`
+- OIDC handler: `packages/*/src/login/oidc/oidcHandlers/AuthorizationCodeWithPkceOidcHandler.ts`
+- Handle incoming redirect: `packages/*/src/login/oidc/redirectHandler/*Handler.ts`,
+the specific Handler depends on which Handler's `canHandle()` method first returns
+`true`.
+
+#### Refresh flow
+
+![Module dependencies](./documentation/diagrams/refresh_flow.svg)
+
+- OIDC handler: `packages/*/src/login/oidc/oidcHandlers/RefreshTokenOidcHandler.ts`
+
+Note that in this case, no redirection happens (i.e., there's only a Backchannel
+exchange): the Access Token is received directly in response to a request
+containing the Refresh Token.
+
+### Helpful resources
+
+Here are some recommended resources to help understand OAuth/OIDC in general:
+- [OAuth 2 in Action](https://www.manning.com/books/oauth-2-in-action), by Justin Richer and Antonio Sanso
+- [OAuth masterclass](https://www.youtube.com/watch?v=egfyV2NV9Mw), by Justin Richer
+- [How To Securely Implement Authentication in Single Page Applications](https://betterprogramming.pub/how-to-securely-implement-authentication-in-single-page-applications-670534da746f), by Dennis Stötzel
+
+These resources may help in getting a better understanding of the authentication
+flows described in the previous section, and in particular by providing relevant
+use cases for each flow.
+
+## Codemap of the client library modules
+
+This section will give a high-level description of the shared inner workings of 
+`@inrupt/solid-client-authn-node` and `@inrupt/solid-client-authn-browser`,
+omitting anything too module-specific.
+
+### The API
+
+Most of the code for these modules is internal and hidden from the developer. The
+public API is located in `packages/*/src/Session.ts`. Developers are expected to
+instantiate a `Session` object, and to use it to interact with the session.
+Usage examples can be found in our
+[public documentation](https://docs.inrupt.com/developer-tools/javascript/client-libraries/tutorial/authenticate/).
+
+### The Handler pattern
+
+Various components, such as the Login and Incoming Redirect components, are based
+on the Handler design pattern. Given data contained in a request, a set of classes
+implementing a similar API will declare whether or not they may handle said request.
+
+- **Handlers** declare two functions:
+  - `canHandle(request)`
+  - `handle(request)`
+  
+  `canHandle()` simply returns a boolean indicating the ability of the handler to handle the 
+  request, and `handle()` actually processes it.
+  
+- All the handlers for a given type of request are tracked by a **Handler Aggregator**, 
```

Comment:
Maybe flip the sentence to start with Handler Aggregator

- **Handler Aggregator** tracks all the handlers for a given type of request. The handler aggregator invokes the first handler for which ...